### PR TITLE
feat(tax): migration file

### PR DESCRIPTION
**Pull request by srindom**
**What**
- Tax Module Migration file.
- Skeleton for API routes and integrations tests for tax API in v2

### CODE REVIEW

File: packages/tax/src/migrations/Migration20240227090331.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    )
+    this.addSql(
+      `CREATE INDEX IF NOT EXISTS "IDX_tax_region_deleted_at" ON "tax_region" ("deleted_at") WHERE deleted_at IS NOT NULL;`
+    )
+    this.addSql(
+      `CREATE UNIQUE INDEX IF NOT EXISTS "IDX_tax_region_unique_country_province" ON "tax_region" ("country_code", "province_code");`
+    )
+
+    // Old foreign key on region_id
+    this.addSql(
```
Comment (srindom) [inline packages/tax/src/migrations/Migration20240227090331.ts:31]: was in doubt about what to do here

File: integration-tests/plugins/__tests__/tax/admin/tax.spec.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    await shutdownServer()
+  })
+
+  afterEach(async () => {
+    const db = useDb()
+    await db.teardown()
+  })
+
+  it("can retrieve a tax rate", async () => {
+    await createAdminUser(dbConnection, adminHeaders)
```
Comment (sradevski) [inline integration-tests/plugins/__tests__/tax/admin/tax.spec.ts]: We can move this to a `beforeEach` maybe?

File: integration-tests/plugins/__tests__/tax/admin/tax.spec.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+describe("Taxes - Admin", () => {
+  let dbConnection
+  let appContainer
+  let shutdownServer
+  let service: ITaxModuleService
+
+  beforeAll(async () => {
+    const cwd = path.resolve(path.join(__dirname, "..", "..", ".."))
+    dbConnection = await initDb({ cwd, env } as any)
+    try {
```
Comment (sradevski) [inline integration-tests/plugins/__tests__/tax/admin/tax.spec.ts]: I think it's fine to just let it explode if it throws here :)

File: packages/medusa/src/api-v2/admin/tax-rates/middlewares.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import { transformQuery } from "../../../api/middlewares"
+
+import { MiddlewareRoute } from "../../../loaders/helpers/routing/types"
+import { authenticate } from "../../../utils/authenticate-middleware"
+
+export const adminTaxRateRoutesMiddlewares: MiddlewareRoute[] = [
+  {
+    method: ["ALL"],
+    matcher: "/admin/tax-rates*",
+    middlewares: [authenticate("admin", ["bearer", "session"])],
```
Comment (sradevski) [inline packages/medusa/src/api-v2/admin/tax-rates/middlewares.ts]: Let's add `api-key` as an auth type as well

File: packages/tax/src/joiner-config.ts [hunk 5]
```diff
@@ -21,5 +25,33 @@ export const joinerConfig: ModuleJoinerConfig = {
...
+      name: ["tax_rate_rule", "tax_rate_rules"],
+      args: {
+        entity: TaxRateRule.name,
+        methodSuffix: "TaxRateRules",
+      },
+    },
+    {
+      name: ["tax_provider", "tax_providers"],
+      args: {
+        entity: TaxRateRule.name,
```
Comment (sradevski) [inline packages/tax/src/joiner-config.ts]: Should this be `TaxProvider.name`?

**Pull request merged**